### PR TITLE
test: cover split-brain during promote

### LINKED ISSUE #9943: Cover with tests, that promote of the previous term is detected and causes Split Brain
Picodata makes a commit, which nopifies PROMOTEs and DEMOTEs in some cases: [link](https://git.picodata.io/picodata/tarantool/-/merge_requests/160/diffs?commit_id=350ebf48cd2b26105f0cf199452c37d95a0f72d7). However, none of a tarantool tests fails, which means that this is not covered with tests. We should fix that.

### PR DESCRIPTION
**Pull request by Serpentian**
This test checks, that when PROMOTE from the prevous term is encountered we immediately notice split-brain situation and break replication without corrupting data.

Closes #9943

NO_DOC=test
NO_CHANGELOG=test

### CODE REVIEW

File: test/replication-luatest/gh_9943_promote_split_brain_test.lua [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+    promote(g.rs.servers[2])
+    g.rs.servers[1]:exec(function(id)
+        -- Without this wait PROMOTE may be not executed and exit with TimedOut
+        -- error. Let's make sure, that server_1 known about new leader.
+        t.helpers.retrying({}, function()
+            t.assert_equals(box.info.election.leader, id)
+        end)
+    end, {g.rs.servers[2]:get_instance_id()})
+    promote(g.rs.servers[1])
```
Comment (Gerold103) [inline test/replication-luatest/gh_9943_promote_split_brain_test.lua:76]: Why did you need to promote 2 and then 1? Could you simply promote 1 right away? If this is to bump the term, then lets add a comment about that.

File: test/replication-luatest/gh_9943_promote_split_brain_test.lua [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    -- Split-Brain should be noticed.
+    set_replication(g.rs.servers[1], {
+        g.rs.servers[1].net_box_uri,
+        g.rs.servers[2].net_box_uri,
+        g.rs.servers[3].net_box_uri,
+    })
+
+    t.helpers.retrying({}, function()
+        t.assert_not_equals(g.rs.servers[1]:grep_log('ER_SPLIT_BRAIN'), nil)
+    end)
```
Comment (Gerold103) [inline test/replication-luatest/gh_9943_promote_split_brain_test.lua:101]: Can you instead get it from the `info.replication` from upstream/downstream status? Log greps is the last resort.

**Review by Gerold103 — COMMENTED** [review]
Thanks for the patch!

File: test/replication-luatest/gh_9943_promote_split_brain_test.lua (hunk 1 above)
Comment (Serpentian) [reply test/replication-luatest/gh_9943_promote_split_brain_test.lua:73]: > If this is to bump the term, then lets add a comment about that

it's, added comments

File: test/replication-luatest/gh_9943_promote_split_brain_test.lua (hunk 2 above)
Comment (Serpentian) [reply test/replication-luatest/gh_9943_promote_split_brain_test.lua]: Sure, it's better to check `box.info.replication.upstream`

**Review by sergepetrenko — COMMENTED** [review]
Thanks for the test!
Please consider adding this case to the `test/replication-luatest/gh_5295_split_brain_test.lua` test. It contains all the similar split-brain test cases. I was sure that the case with an old promote is also handled there, but apparently it's not.

**Review by Gerold103 — APPROVED** [review]
Thanks for the test!

**Comment by Serpentian** [comment]
> Please consider adding this case to the test/replication-luatest/gh_5295_split_brain_test.lua test.

Moved the test there, no changes to the test were done

File: test/replication-luatest/gh_5295_split_brain_test.lua [hunk 3]
```diff
@@ -380,3 +380,99 @@ g_very_old_term.test_confirm = function( cg)
...
+    cg.servers[3]:exec(update_replication, {})
+    cg.servers[1]:exec(update_replication, new_replication)
+    cg.servers[2]:exec(update_replication, new_replication)
+
+    -- Bump term on the first server, so that it is more, than term
+    -- of the third server after its promotion.
+    promote(cg.servers[2])
+    cg.servers[1]:exec(function(id)
+        -- Without this wait PROMOTE may be not executed and exit with TimedOut
+        -- error. Let's make sure, that server_1 known about new leader.
```
Comment (sergepetrenko) [inline test/replication-luatest/gh_5295_split_brain_test.lua:418]: typo: known->knows

File: test/replication-luatest/gh_5295_split_brain_test.lua [hunk 4]
```diff
@@ -380,3 +380,99 @@ g_very_old_term.test_confirm = function( cg)
...
+        t.helpers.retrying({}, function()
+            t.assert_equals(box.info.election.leader, id)
+        end)
+    end, {cg.servers[2]:get_instance_id()})
+    promote(cg.servers[1])
```
Comment (sergepetrenko) [inline test/replication-luatest/gh_5295_split_brain_test.lua:423]: Can you reuse the `write_promote` function defined above?

File: test/replication-luatest/gh_9943_promote_split_brain_test.lua (hunk 2 above)
Comment (sergepetrenko) [reply test/replication-luatest/gh_9943_promote_split_brain_test.lua]: Resolving since @Gerold103 approved the PR.

**Review by sergepetrenko — COMMENTED** [review]
Thanks for the fixes!

File: test/replication-luatest/gh_5295_split_brain_test.lua (hunk 3 above)
Comment (Serpentian) [reply test/replication-luatest/gh_5295_split_brain_test.lua]: Dropped the comment. Since now `write_promote` is used, I sync severs by their limbo term, which is enough to stabilize the test

File: test/replication-luatest/gh_5295_split_brain_test.lua (hunk 4 above)
Comment (Serpentian) [reply test/replication-luatest/gh_5295_split_brain_test.lua]: Sure, done

**Review by sergepetrenko — APPROVED** [review]
Thanks for the test!
LGTM.

**Comment by sergepetrenko** [comment]
Backports:
3.1 https://github.com/tarantool/tarantool/pull/10255
2.11 https://github.com/tarantool/tarantool/pull/10254
